torch/_dynamo/guards.py

```diff
@@ -410,36 +395,31 @@ def TENSOR_MATCH(self, guard: Guard):
         if guard.is_nn_module():
             self.ID_MATCH(guard)
         else:
+            self.TYPE_MATCH(guard)
             value = self.get(guard.name)
             assert isinstance(value, torch.Tensor)
             tensor_name = self.arg_ref(guard)
-            # [Note - On Export Tensor Guards]
-            #
-            # In eager mode, tensor guards are evaluated through C++, in guards.cpp
-            # see [Note - On Eager Tensor Guards] for more info.
+            # [Note - On Tensor Guards]
             #
             # In export mode, we instead maintain parallel logic between C++ and python
             # here, with an exception of checking the dispatch key - with the idea that a dispatch key
             # is an entirely runtime notion that would make no sense to keep in an exported graph.
             #
             # The list of tensor fields and calls we care about can be found in `terms` below.
-            if self.check_fn_manager.output_graph.export:
-                # An interesting observation, is that the code below
-                # + a dispatch_key check + TYPE_MATCH performs almost identically
-                # to our C++ guards on a small sweep.
-                # TODO(voz) investigate removing C++ guards
-                terms = ["dtype", "device.index", "requires_grad", "ndimension()"]
-                if not config.dynamic_shapes:
-                    terms.extend(["size()", "stride()"])
-
-                code = []
-                for term in terms:
-                    real_value = self.get(tensor_name + "." + term)
-                    code.append(f"{tensor_name}.{term} == {real_value}")
-                self._produce_guard_code(guard, code)
-            else:
-                self.tensor_check_names.append(tensor_name)
-                self.tensor_check_examples.append(value)
+            terms = ["dtype", "device.index", "requires_grad", "ndimension()"]
+            if not config.dynamic_shapes:
+                terms.extend(["size()", "stride()"])
+
+            code = []
+            # TODO(voz): Ask ed how to turn a keyset like DispatchKeySet(CUDA, ADInplaceOrView, AutogradCUDA, AutocastCUDA) into a
```

Comment:
@ezyang help

Reply:
There's a crappy binding at `_dispatch_keys` in `torch/csrc/utils/python_dispatch.cpp` but it is probably not very fast.